Add Kafka Json encoder

charlesjmorgan opened the pull request:
Add `JsonRowEncoder` and `JsonRowEncoderFactory`
Add test case in `io.prestosql.plugin.kafka.TestKafkaIntegrationSmokeTest#testRoundTripAllFormats`

Might it be better to have separate formatters for the date/time types rather than methods in `JsonRowEncoder`?

closes #3980

aalbu commented on presto-kafka/src/test/java/io/prestosql/plugin/kafka/TestKafkaIntegrationSmokeTest.java:
```diff
@@ -160,6 +162,119 @@ private KafkaTopicFieldDescription createOneFieldDescription(String name, Type t
+                                        "TIMESTAMP '2020-07-15 01:02:03.456 UTC'"))))
+                .add(new JsonDateTimeTestCase(
+                        "seconds_since_epoch",
+                        ImmutableList.of("f_time", "f_time_with_time_zone", "f_timestamp", "f_timestamp_with_time_zone"),
```
Since the column list is the same for each test case, you can hard-code it in the test.

charlesjmorgan replied:
unfortunately seconds/milliseconds since epoch don't support date

aalbu commented on presto-kafka/src/test/java/io/prestosql/plugin/kafka/TestKafkaIntegrationSmokeTest.java:
```diff
@@ -160,6 +162,119 @@ private KafkaTopicFieldDescription createOneFieldDescription(String name, Type t
+        {
+            String[] values = new String[rowValues.size()];
+            for (int i = 0; i < rowValues.size(); i++) {
+                values[i] = "(" + rowValues.get(i).stream().map(Object::toString).collect(Collectors.joining(", ")) + ")";
```
`String::valueOf` would be safer than `Object::toString` here.

aalbu commented on presto-kafka/src/test/java/io/prestosql/plugin/kafka/encoder/json/TestJsonDateTime.java:
```diff
@@ -0,0 +1,199 @@
+
+        RowEncoder rowEncoder = createRowEncoder(col1, col2, col3, col4, col5);
+
+        Block dateBlock = DateType.DATE.createBlockBuilder(null, 1).writeInt(18458);
```
TBH, I find this test difficult to read.  I am looking at the setup and I wouldn't be able to predict what the outcome should be (not without writing a small program).  Maybe I'm missing something, but I don't think it adds anything to the tests in `TestKafkaIntegrationSmokeTest`.  If you wanted to see the full JSON message that is generated, you can get that in the round-trip test too (in `_message`).

charlesjmorgan replied:
I wrote these tests for development because they are a million times faster than the roundtrip test in `TestKafkaIntegrationSmokeTest`. They both serve the same purpose now, idk if it would be a good idea to keep them around for development or if I should just use the roundtrip tests.

aalbu commented on presto-kafka/src/test/java/io/prestosql/plugin/kafka/TestKafkaIntegrationSmokeTest.java:
```diff
@@ -189,10 +304,14 @@ public void testRoundTripAllFormats(RoundTripTestCase testCase)
+                .add(new RoundTripTestCase(
+                        "all_datatypes_json",
+                        ImmutableList.of("f_bigint", "f_int", "f_smallint", "f_tinyint", "f_double", "f_boolean", "f_varchar"),
+                        ImmutableList.of(100000, 1000, 100, 10, 1000.001, true, "'test'")))
```
I think there was a suggestion in a previous review about testing with more than one record.

charlesjmorgan replied:
yes, that commit is in the raw encoder pr #4417. I'll make sure that get's brought in when it merges

losipiuk commented on presto-kafka/src/test/java/io/prestosql/plugin/kafka/TestKafkaIntegrationSmokeTest.java:
```diff
@@ -160,6 +162,119 @@ private KafkaTopicFieldDescription createOneFieldDescription(String name, Type t
     }
 
+    @Test(dataProvider = "testJsonDateTimeFormatsDataProvider")
+    public void testJsonDateTimeFormats(JsonDateTimeTestCase testCase)
```
how is logic here different from `testRoundTripAllFormats`?
Can we make `testRoundTripAllFormats` and `testJsonDateTimeFormats` share testcase definition.

charlesjmorgan replied:
yes, that's the plan when the raw encoder pr merges #4417

charlesjmorgan commented on presto-kafka/src/main/java/io/prestosql/plugin/kafka/encoder/json/JsonRowEncoder.java:
```diff
@@ -0,0 +1,382 @@
+                checkArgument(columnHandle.getDataFormat() == null, "Unexpected data format '%s' defined for column '%s'", columnHandle.getDataFormat(), columnHandle.getName());
+            }
+            else {
+                checkArgument(DATE_TIME_FORMATS.contains(columnHandle.getDataFormat()), "Incorrect data format '%s' defined for column '%s'", columnHandle.getDataFormat(), columnHandle.getName());
```
yes, that would be a good idea

charlesjmorgan commented on presto-kafka/src/main/java/io/prestosql/plugin/kafka/encoder/json/JsonRowEncoder.java:
```diff
@@ -0,0 +1,382 @@
+    private String formatDateTime(EncoderColumnHandle columnHandle, SqlDate value)
+    {
+        switch (columnHandle.getDataFormat()) {
+            case "custom-date-time":
```
oh an enum would be good. What do you think of having separate classes for the formatters? We could precompile a list of formatters at encoder initialization.

charlesjmorgan commented:
Implemented all the changes in the resolved review comments, will make it through the rest on Monday. I changed the date/time formatting to precompile a list of functions to format the values, lmk what you think of this.

losipiuk commented on presto-kafka/src/main/java/io/prestosql/plugin/kafka/encoder/json/format/DefaultFormatFunction.java:
```diff
@@ -0,0 +1,20 @@
+ */
+package io.prestosql.plugin.kafka.encoder.json.format;
+
+public interface DefaultFormatFunction
```
this one is never used in practice, as whenever you take something out of `formatFunctions` you cast it to `JsonDateTimeFormatFunction`. Is that a bug?
Also I do not huge fan of idea of having format functions which do not satisfy common interface in single list.

charlesjmorgan replied:
I needed a filler function for non-temporal columns. I agree this is less than ideal, I'm not sure what to replace it with. This function is never actually used in the encoder. I was thinking that in the future we could replace all the `appendLong, appendBoolean...` methods with a precompiled list of lambda functions to format/add the column value to the encoder. For now we need to figure out a different solution. I could change the `List<JsonFormatFunctions> formatFunctions` field in `JsonRowEncoder` to a `Map` and have the column name as the key? That way I don't need to access it with the `currentColumnIndex`.

losipiuk replied:
I find the approach with default function cumbersome. Maybe instead having a hierarchy here we can just have as set of functions and single class with methods delegating to functors. Then we can provide the fuctions we care about as lambdas via builder. The ones which are not explicitly set would just throw exception. 

Sth like this:
```java
public class JsonFormatFunction
{
    private final Function<Instant, String> formatTimestampLegacyFunc;
    private final Function<LocalDateTime, String> formatTimestampFunc;
    //private final Function<Something, String> formatSomething;

    public JsonFormatFunction(Function<Instant, String> formatTimestampLegacyFunc, Function<LocalDateTime, String> formatTimestampFunc)
    {
        this.formatTimestampLegacyFunc = formatTimestampLegacyFunc;
        this.formatTimestampFunc = formatTimestampFunc;
    }

    public String formatTimestampLegacy(Instant value)
    {
        return formatTimestampLegacyFunc.apply(value);
    };

    public String formatTimestamp(LocalDateTime value)
    {
        return formatTimestampFunc.apply(value);
    }

    public static class Builder
    {
        private Function<Instant, String> formatTimestampLegacyFunc = (ignored) -> { throw new RuntimeException("not supported argument type"); };
        private Function<LocalDateTime, String> formatTimestampFunc = (ignored) -> { throw new RuntimeException("not supported argument type"); };

        JsonFormatFunction build()
        {
            return new JsonFormatFunction(
                    formatTimestampLegacyFunc,
                    formatTimestampFunc);
        }

        public Builder setFormatTimestampLegacyFunc(Function<Instant, String> formatTimestampLegacyFunc)
        {
            this.formatTimestampLegacyFunc = formatTimestampLegacyFunc;
            return this;
        }

        public Builder setFormatTimestampFunc(Function<LocalDateTime, String> formatTimestampFunc)
        {
            this.formatTimestampFunc = formatTimestampFunc;
            return this;
        }
    }
}
```

One downside of this is that is a bit verbose. It also involves extra boxing/unboxing - not sure how much of a problem that is with new JVMs.

losipiuk commented on presto-kafka/src/main/java/io/prestosql/plugin/kafka/encoder/json/JsonRowEncoder.java:
```diff
@@ -0,0 +1,267 @@
+    @Override
+    protected void appendSqlTimestamp(SqlTimestamp value)
+    {
+        if (session.isLegacyTimestamp()) {
```
This looks wrong. It does not seem to me that you should not use same formatting function and just pass different version to it based on selected timestamp semantic.
As the semantics is different the formatting should be different. I.e it does not make sense to append any timezone information for non-legacy semantics in output string.

It is hard for me to wrap my head around it though.

charlesjmorgan replied:
Ik the timestamp changes have me confused as well (clearly hahah). I think I've been testing with the legacy timestamp, which would explain why I haven't encountered a problem with this yet, so I'll add different functions and test with the new timestamp.

losipiuk replied:
It seems to me that cleanest approach would be to not allow encoding timestamp in non-legacy semantic to any format which clearly denotes an instant (i.e. explicitly specifies a timezone in a value). The problem with that is it would be not compatible with what we already have in decoders (which do not care about timestamp semantics whatsoever). 

Maybe we should assume that logic in decoders only makes sense for legacy timestamp semantics, and enforce that with check in the connector. Then we can focus on implementing just legacy semantics of timestamp in encoders in a way that is compatible with decoders.

Then we can add non-legacy semantics to both decoders and encoders?

The problem with that is that we would break people which use kafka timestamp decoders with timestamp semantic set to non-legacy. I do not expect that it would be many users, but one never knows.

@findepi, as an expert on hard issues, can you chime in here?

findepi replied:
@losipiuk i like your proposal.
i would only add that there seems to be no fundamental problem in decoding `timestamp` (without zone) from a value containing a zone. You could think of it as a CAST.
Eg JDBC ResultSet.getDate can be called on a date or timestamp, and returns a date part of the value.

I am OK with dropping this though, for non-legacy. We can always re-add this if we learn about particular requirements.

losipiuk replied:
Thanks for comment.

> @losipiuk i like your proposal.
> i would only add that there seems to be no fundamental problem in decoding `timestamp` (without zone) from a value containing a zone. You could think of it as a CAST.
> Eg JDBC ResultSet.getDate can be called on a date or timestamp, and returns a date part of the value.
> 
> I am OK with dropping this though, for non-legacy. We can always re-add this if we learn about particular requirements.

Yeah - it makes sense for decoding. Less so for encoding, as it not obvious what timezone use in encoded value. And even if timezone is statically defined by encoding format it is kinda smelly.

losipiuk commented on presto-kafka/src/test/java/io/prestosql/plugin/kafka/TestKafkaIntegrationSmokeTest.java:
```diff
@@ -160,7 +162,120 @@ private KafkaTopicFieldDescription createOneFieldDescription(String name, Type t
+
+    private List<JsonDateTimeTestCase> jsonDateTimeFormatsData()
+    {
+        return ImmutableList.<JsonDateTimeTestCase>builder()
```
I think those test would benefit if stracture it more granurally, where single testcase defines 
 * format
 * column
 * expected value

Sth like this:
```java
        testCaseBuilder.
            .addTestCase("custom_date_time_json", "f_time", "TIME '01:02:03.456'")
            .addTestCase("custom_date_time_json", "f_time", "TIME '03:02:03.456'")
            .addTestCase("custom_date_time_json", "f_time_with_time_zone", "TIME '05:02:03.456 UTC'")
            .addTestCase("iso8601_json", "f_time", "TIME '05:02:03.456'")
```

I think it is easier to read as you have all information about single testcase in single line.
We can also focus more on data types which are more problematic and have more corner cases.

Under the hood we can insert whole row with all columns but one we care about set to null.

WDYT?

losipiuk replied:
On the other hand maybe the unit tests we have in `TestJsonDateTime` are enough and these should be treated as just smoke coverage?
@aalbu ?

charlesjmorgan replied:
I agree this format would be much easier to read and debug. Could probably just use one table as well.

aalbu replied:
@losipiuk I think lighter coverage here and testing all edge cases in `TestJsonDateTime` is a good idea.  Perhaps what we would need to add here would be some round-trip tests with legacy semantics enabled.  We should talk about that.

charlesjmorgan replied:
I think that it would be better to test each individual format function. The current tests in `TestJsonDateTime` helped me figure out how to write the format functions, but they just aren't great tests.

charlesjmorgan replied:
+1 to lighter smoke testing

losipiuk replied:
> we should talk about that

@aalbu see https://github.com/prestosql/presto/pull/4477#discussion_r458737542

losipiuk commented on presto-kafka/src/test/java/io/prestosql/plugin/kafka/encoder/json/TestJsonDateTime.java:
```diff
@@ -0,0 +1,195 @@
+        testDateHelper(18458, "2020-07-15", "iso8601", Optional.empty());
+        testTimeHelper(1594846503123L, "20:55:03.123", "iso8601", Optional.empty());
+        testTimeWithTZHelper(packDateTimeWithZone(1594846503123L, "UTC"), "20:55:03.123Z", "iso8601", Optional.empty());
+        testTimestampHelper(1594846503123L, "2020-07-15T20:55:03.123", "iso8601", Optional.empty());
```
Use of datatype internal representation here make the test look cryptic. Compute `1594846503123L` using `java.time.*` instead hardcoding it in tests?

aalbu commented on presto-kafka/src/main/java/io/prestosql/plugin/kafka/encoder/json/format/CustomDateTimeFormatFunctions.java:
```diff
@@ -0,0 +1,85 @@
+    {
+        try {
+            DateTimeFormatter formatter = DateTimeFormat.forPattern(formatHint).withLocale(Locale.ENGLISH);
+            return (millis -> (new DateTime(millis, DateTimeZone.UTC)).toString(formatter));
```
The parentheses around the return value are not necessary (and I personally feel they reduce readability).  I also like more passing the temporal value to the formatter:
`return millis -> formatter.print(new DateTime(millis, UTC));`

A possible optimization could be pre-instantiating formatters for all patterns when creating the encoder and reusing them (the Joda ones are thread safe), as opposed to doing it for every row.

charlesjmorgan replied:
Shouldn't these functions reuse the same `DateTimeFormatter's`?

aalbu replied:
You're right, you're actually doing what I was suggesting.  I misread the code.

charlesjmorgan commented:
Changes:
- Restructure json date/time formatters according to this comment [https://github.com/prestosql/presto/pull/4477#discussion_r458728707](url)
- Format functions return a parameterized `Function`
- Add unit tests for format functions
- Restructure Json Date/Time round trip tests in `TestKafkaIntegrationSmokeTest` for readability

charlesjmorgan commented:
I think I made the correct suggested legacy timestamp changes. The legacy format functions definitely need a close look at tho

charlesjmorgan commented on presto-kafka/src/main/java/io/prestosql/plugin/kafka/encoder/json/format/JsonFormatFunctions.java:
```diff
@@ -0,0 +1,195 @@
+        }
+    }
+
+    public static Function<Long, String> timestampFormatFunction(FormatType formatType, Optional<String> formatHint)
```
removed and replaced with legacy function

charlesjmorgan commented on presto-kafka/src/main/java/io/prestosql/plugin/kafka/encoder/json/format/JsonFormatFunctions.java:
```diff
@@ -0,0 +1,195 @@
+        }
+    }
+
+    public static Function<Long, String> legacyTimestampFormatFunction(FormatType formatType, Optional<String> formatHint)
```
tried to rename functions to be more descriptive

aalbu commented on presto-kafka/src/main/java/io/prestosql/plugin/kafka/encoder/json/format/CustomDateTimeFormatFunctions.java:
```diff
@@ -0,0 +1,91 @@
+    public static Function<Long, String> formatDateFunc(String formatHint)
+    {
+        try {
+            DateTimeFormatter formatter = DateTimeFormat.forPattern(formatHint).withLocale(Locale.ENGLISH).withZoneUTC();
```
You can use a static import (below too).

findepi replied:
is the format hint documented to be Joda Time?

aalbu commented on presto-kafka/src/main/java/io/prestosql/plugin/kafka/encoder/json/format/CustomDateTimeFormatFunctions.java:
```diff
@@ -0,0 +1,91 @@
+    {
+        try {
+            DateTimeFormatter formatter = DateTimeFormat.forPattern(formatHint).withLocale(Locale.ENGLISH).withZoneUTC();
+            return millis -> (new DateTime(millis, DateTimeZone.UTC)).toString(formatter);
```
Let's be consistent in the way we format.  Either `new DateTime(...).toString(fromatter)` as here, or `formatter.print(new DateTime(...))` as below.

findepi replied:
I think it would be better to supplement the session zone here.
This would make it more similar to the non-legacy case, when we implement it.

Ie 
`INSERT ,.. TIMESTAMP 'some_date_time'`
would be `some_date_time session_zone` in legacy
and `some_date_time` in non-legacy

and if the format hint does not contain the zone,
they would actually be the same.

aalbu commented on presto-kafka/src/main/java/io/prestosql/plugin/kafka/encoder/json/format/JsonFormatFunction.java:
```diff
@@ -0,0 +1,118 @@
+
+    public static class Builder
+    {
+        private Function<Long, String> formatDateFunc = (ignored) -> { throw new RuntimeException("unsupported argument type"); };
```
You could define
```
private static final Function<Long, String> UNIMPLEMENTED = ignored -> { throw new RuntimeException("unsupported argument type"); };
```
and reference that:
```
private Function<Long, String> formatDateFunc = UNIMPLEMENTED;
private Function<Long, String> formatTimeFunc = UNIMPLEMENTED;
...
```

aalbu commented on presto-kafka/src/main/java/io/prestosql/plugin/kafka/encoder/json/format/JsonFormatFunctions.java:
```diff
@@ -0,0 +1,172 @@
+
+    public static long daysToEpochMillis(long value)
+    {
+        return DAYS.toMillis(value);
```
This is subjective, but you could just as well inline.  I think `DAYS.toMillis(value)` reads pretty well.

aalbu commented on presto-kafka/src/main/java/io/prestosql/plugin/kafka/encoder/json/format/MillisecondsFormatFunctions.java:
```diff
@@ -0,0 +1,33 @@
+
+    public static Function<Long, String> formatMillisWithTZFunc()
+    {
+        return encodedMillisWithTZ -> String.valueOf(unpackMillisUtc(encodedMillisWithTZ));
```
Perhaps we should not support this scenario.  This does not seem a meaningful return value - without the tz info, the milliseconds don't make much sense here, right?

aalbu replied:
Sorry, that was wrong.  This format implies millis since epoch.

aalbu commented on presto-kafka/src/main/java/io/prestosql/plugin/kafka/encoder/json/format/RFC2822FormatFunctions.java:
```diff
@@ -0,0 +1,89 @@
+                        unpackZoneKey(encodedMillisWithTZ).getZoneId()));
+    }
+
+    private static LocalDateTime localDateTimeOfEpochMillis(long epochMillis)
```
Perhaps this would warrant being public in `JsonFormatFunctions`.  It's being used in `ISO8601FormatFunctions`, as well.

aalbu commented on presto-kafka/src/main/java/io/prestosql/plugin/kafka/encoder/json/format/SecondsFormatFunctions.java:
```diff
@@ -0,0 +1,34 @@
+        return millis -> String.valueOf(millisToSeconds(millis));
+    }
+
+    public static Function<Long, String> formatSecondsWithTZFunc()
```
Another situation in which we're throwing away tz info.

findepi replied:
This is like unix `time()`, it's fine.

findepi commented on presto-kafka/src/main/java/io/prestosql/plugin/kafka/encoder/json/JsonRowEncoder.java:
```diff
@@ -0,0 +1,266 @@
+            checkArgument(isSupportedType(columnHandle.getType()), "Unsupported column type '%s' for column '%s'", columnHandle.getType(), columnHandle.getName());
+
+            if (isDateTimeType(columnHandle.getType())) {
+                checkArgument(columnHandle.getDataFormat() != null, "Unsupported or no dataFormat '%s' defined for temporal column '%s'", columnHandle.getDataFormat(), columnHandle.getName());
```
```suggestion
                checkArgument(columnHandle.getDataFormat() != null, "No dataFormat defined for temporal column '%s'", columnHandle.getName());
```

findepi commented on presto-kafka/src/main/java/io/prestosql/plugin/kafka/encoder/json/JsonRowEncoder.java:
```diff
@@ -0,0 +1,266 @@
+        else if (type == TIME) {
+            return JsonFormatFunction.builder().setFormatTimeFunc(getTimeFormatFunction(dataFormat, formatHint)).build();
+        }
+        else if (type == TIME_WITH_TIME_ZONE) {
```
redundant `else` (https://github.com/prestosql/presto/#additional-ide-configuration)

findepi commented on presto-kafka/src/main/java/io/prestosql/plugin/kafka/encoder/json/format/CustomDateTimeFormatFunctions.java:
```diff
@@ -0,0 +1,91 @@
+{
+    private CustomDateTimeFormatFunctions() {}
+
+    public static Function<Long, String> formatDateFunc(String formatHint)
```
avoid abbreviations

func -> function

findepi commented on presto-kafka/src/main/java/io/prestosql/plugin/kafka/encoder/json/format/CustomDateTimeFormatFunctions.java:
```diff
@@ -0,0 +1,91 @@
+            DateTimeFormatter formatter = DateTimeFormat.forPattern(formatHint).withLocale(Locale.ENGLISH).withZoneUTC();
+            return millis -> (new DateTime(millis, DateTimeZone.UTC)).toString(formatter);
+        }
+        catch (IllegalArgumentException e) {
```
Code is correct but I would find it more readable if the lambda was outside of try block...

findepi commented on presto-kafka/src/main/java/io/prestosql/plugin/kafka/encoder/json/format/CustomDateTimeFormatFunctions.java:
```diff
@@ -0,0 +1,91 @@
+    {
+        try {
+            DateTimeFormatter formatter = DateTimeFormat.forPattern(formatHint).withLocale(Locale.ENGLISH).withZoneUTC();
+            return encodedMillisWithTZ -> formatter.print(new DateTime(
```
encodedMillisWithTZ -> value

findepi commented on presto-kafka/src/main/java/io/prestosql/plugin/kafka/encoder/json/format/CustomDateTimeFormatFunctions.java:
```diff
@@ -0,0 +1,91 @@
+                    DateTimeZone.forID(unpackZoneKey(encodedMillisWithTZ).getId())));
+        }
+        catch (IllegalArgumentException e) {
+            throw new IllegalArgumentException(format("Invalid joda pattern '%s' passed as format hint", formatHint), e);
```
joda -> Joda Time

findepi commented on presto-kafka/src/main/java/io/prestosql/plugin/kafka/encoder/json/format/JsonFormatFunction.java:
```diff
@@ -0,0 +1,118 @@
+        return formatTimeFunc.apply(value);
+    }
+
+    public String formatTimeWithTZ(long value)
```
withTZ -> withTimeZone

charlesjmorgan commented:
removed temporal column support from this pr, will open another pr for that

findepi commented:
Merged, thanks!
